Add draft application models and behaviour

casassg opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines  https://github.com/HackAssistant/registration/blob/master/.github/CONTRIBUTING.md
-->

## What this PR does / Why we need it
Allow hackers to save applications half way. This should enable to have also fields pre-filled if we add the myMLH sign-up option. It also helps make the burden of filling the application easier. 

To find hackers that have forgotten to submit we can probably add a SQL query to pull all hackers who have not finished their applications.

This could possibly backfire. It could be possible to have more hackers that forget to finish their application. 

Would appreciate any feedback on this. We can also merge without the option of saving and use it for the MyMLH integration.

casassg commented:
I may try to do this by using Django cache instead. Maybe better for this, avoids having to create yet another table in the database.

qaisjp commented:
> This could possibly backfire. It could be possible to have more hackers that forget to finish their application.

How about auto submitting all partially filled applications when the deadline hits? (or more accurately/safely — allowing the application review process also process partially filled applications)

casassg commented:
The problem is that partially submitted applications can't be added to the database as it is right now, some null constraints will not be followed.

 Another option is to have the saving process happen in the background (every minute or 2) and show it in the bottom: `Saving...`. It would be similar to how medium.com works. Recognizes when you have finished writing something and saves it.

dasilvacontin commented:
Hey! In order to see this form, you need to be logged in right?

LGTM to have it stored, after a number of changes are done, or whatever. If you worry about people forgetting that they started but didn't submit, you can always send an email reminder :)

casassg commented:
You do need to be logged in.

I may add a document for good practices of users that have not finished applying, but I think it would work. My main worry was due to the fact that I added another button next to the submit application and that may confuse people.
![screencapture-localhost-8000-dashboard-2018-11-15-02_11_35](https://user-images.githubusercontent.com/6912589/48542208-cd6e9d00-e87b-11e8-9f33-beb992d0fb20.png)
 I plan on removing the button and doing save automatically every time the user focuses out of a field or stops writing. Any recs on doing this @dasilvacontin ?

peasant98 commented:
Yeah, we could send a friendly email when the deadline's really approaching, to remind hackers to finish and submit their partial application when they haven't finished.

casassg commented:
Moved to use background requests instead. After 5 seconds of inactivity, the browser sends an AJAX POST request with all the data to the `save_draft` endpoint.

Removed the `Save progress` button. This should avoid unnecessary confusion.

casassg commented:
Added documentation and added a signal to delete DraftApplication object from DB once the application has been submitted.

carlcortright commented on applications/signals.py:
```diff
@@ -0,0 +1,13 @@
+
+# Delete DraftApplication when application submitted
+@receiver(post_save, sender=models.Application)
+def clean_draftapplication(sender, instance, created, *args, **kwargs):
```
Can we rename this to `clean_draft_application`?
```suggestion
def clean_draft_application(sender, instance, created, *args, **kwargs):
```

carlcortright replied:
Feel free to check me on this, I know it's also a model name

carlcortright commented on applications/apps.py:
```diff
@@ -0,0 +1,12 @@
+    def ready(self):
+        super(ApplicationsConfig, self).ready()
+        from .signals import clean_draftapplication
+        clean_draftapplication
```
see below about `clean_draft_application` naming
```suggestion
        clean_draft_application
```

carlcortright commented on applications/apps.py:
```diff
@@ -0,0 +1,12 @@
+
+    def ready(self):
+        super(ApplicationsConfig, self).ready()
+        from .signals import clean_draftapplication
```
```suggestion
        from .signals import clean_draft_application
```

carlcortright replied:
Small nit, I think it's more in-line with the rest of the codebase.

carlcortright commented on README.md:
```diff
@@ -210,7 +211,26 @@ server {
 - **is_director**: Allows user to send invites to hackers as well as send reimbursement.
 - **is_admin**: Allows user to enter Django Admin interface
 
+### Important SQL queries
```
Thanks for adding these 🙏

carlcortright commented on applications/templates/include/application_form.html:
```diff
@@ -99,6 +99,60 @@ <h3>{{ fieldset.name }}<br>
+                        clearTimeout(clear_timer);
+                    }
+                    $('.saved-string').html('Saving...');
+                    var o = {};
```
nit: can we change this to something more descriptive

carlcortright replied:
Idk what it does otherwise i'd update

carlcortright commented on applications/templates/include/application_form.html:
```diff
@@ -99,6 +99,60 @@ <h3>{{ fieldset.name }}<br>
+                    }
+                    $('.saved-string').html('Saving...');
+                    var o = {};
+                    var a = $('form').serializeArray();
```
^^ same with `a`

carlcortright commented on applications/views.py:
```diff
@@ -178,3 +183,13 @@ def post(self, request, *args, **kwargs):
+
+
+@login_required
+def save_draft(request):
```
Seeing this makes me think someone could attack this by saving other fields such as `STATUS`. I don't think it's possible to send those with the application when the user submits. Wanted to flag just in case.

casassg replied:
Fair point. Will narrow down to fields on model and on form. That should make it safe

casassg replied:
It should still not be able as draftapplication is only used to fill fields in the form between sessions. But it's not too bad to narrow down so that we don't forget about it later on.
